---

# Task 1: Familiarize Yourself with the Dataset

There are two main subtasks:  
1. Load the four data files that are part of this dataset.  
2. Explore each dataset and define the data types of each column. Describe the contents of each column, including the nature of the data values. Follow a similar approach as demonstrated in the lecture and tutorial. Discuss the shape of the data, value ranges, and possible categories. Convert any data into a more appropriate format if needed.  

Add as many code and markdown cells as necessary below. Typically, you should perform one type of investigation per code cell with output, and then comment on the results in a markdown cell below it.  

After completing the analysis, write a concise summary in the **Task 1: summary report** cell at the end. The final report should succinctly summarize the main characteristics of each data file, how they relate to each other, and any notable features.


### **Loading Data Files**

In [1]:
import pandas as pd

In [2]:
sessions = pd.read_csv('sessions.csv')
streamers = pd.read_csv('streamers.csv')
streams = pd.read_csv('streams.csv')
viewers = pd.read_csv('viewers.csv')

### **Dataset Exploration**

**Sessions**

In [10]:
sessions.head()

,session_id,viewer_id,streamer_id,stream_id,started_at,ended_at,duration_mins,chat_messages_sent,bits_cheered,followed_during,subscribed_during
0,SES000001,VWR00938,STR0055,SM00860,2024-01-18 19:16:00,2024-01-18 20:28:00,72,0,0,False,False
1,SES000002,VWR00804,STR0034,SM00542,2024-01-04 17:11:00,2024-01-04 19:41:00,150,11,0,False,False
2,SES000003,VWR00704,STR0008,SM00124,2024-02-15 15:11:00,2024-02-15 16:53:00,102,5,0,False,False
3,SES000004,VWR00926,STR0024,SM00373,2024-01-18 19:59:00,2024-01-18 20:56:00,57,3,9663,False,False
4,SES000005,VWR00883,STR0008,SM00120,2024-03-19 20:08:00,2024-03-19 20:22:00,14,7,0,False,False


Looking at this data file, I can see that each session, viewer, streamer, and stream has its own ID composed of letters and numbers. There are timestamps for the start and end of the stream. Numeric values are used to express the duration in munutes, number of chats sent, and number of cheers (chats sent with bits, virtual currency). I can see that booleans (True or False) are used to tell whether or not people followed during the stream and if they subscribed during the stream.

In [19]:
sessions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22825 entries, 0 to 22824
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   session_id          22825 non-null  object
 1   viewer_id           22825 non-null  object
 2   streamer_id         22765 non-null  object
 3   stream_id           22825 non-null  object
 4   started_at          22825 non-null  object
 5   ended_at            22825 non-null  object
 6   duration_mins       22825 non-null  int64 
 7   chat_messages_sent  22825 non-null  int64 
 8   bits_cheered        22825 non-null  int64 
 9   followed_during     22825 non-null  bool  
 10  subscribed_during   22825 non-null  bool  
dtypes: bool(2), int64(3), object(6)
memory usage: 1.6+ MB


There are a total of 22825 entries with a total of 11 columns. There are no null (missing) values here. The values are a mix of data types, including objects (usually strings), ints, and booleans.

In [23]:
sessions.describe()

,duration_mins,chat_messages_sent,bits_cheered
count,22825.000000,22825.000000,22825.000000
mean,59.051303,6.442629,398.425323
std,60.383979,46.576309,1571.528658
min,0.000000,0.000000,0.000000
25%,15.000000,0.000000,0.000000
50%,39.000000,1.000000,0.000000
75%,83.000000,5.000000,0.000000
max,443.000000,1987.000000,10000.000000


For the numeric data in this file, streams average about 59 minutes with a minimum length of 0 minutes and maximum length of 443 minutes. The average amount of chat messages sent is 6 with a minimum of 0 and maximum of 1987. For cheers, there is an average of 398 with a minimum of 0 and maximum of 10,000. Looking at the entire dataset, it seems like more people are commenting but in smaller amounts of comments. For cheers, there are fewer instances of people sending bits, but when they do, they send a large amount of bits.

**Streamers**

In [11]:
streamers.head()

,streamer_id,streamer_name,category,language,partner_status,total_followers,avg_concurrent_viewers
0,STR0001,xQc,Just Chatting,English,True,11500000,138706
1,STR0002,Pokimane,Just Chatting,English,True,-25551,71660
2,STR0003,HasanAbi,Just Chatting,English,True,2600000,34220
3,STR0004,Amouranth,Just Chatting,English,True,6200000,68607
4,STR0005,Mizkif,Just Chatting,English,True,2100000,6771


This data file focuses on each streamer, who gets a unique ID. The data includes the streamer's name, their streaming category, language, whether or not the streamer is a partner, how many followers they have, and their average number of concurrent viewers. Here there is also a combination of string, boolean, and numeric data.

In [20]:
streamers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 76 entries, 0 to 75
Data columns (total 7 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   streamer_id             76 non-null     object
 1   streamer_name           76 non-null     object
 2   category                76 non-null     object
 3   language                68 non-null     object
 4   partner_status          76 non-null     bool  
 5   total_followers         76 non-null     int64 
 6   avg_concurrent_viewers  76 non-null     int64 
dtypes: bool(1), int64(2), object(4)
memory usage: 3.8+ KB


There are a total of 76 rows of data across 7 columns. There is data in the form of objects (strings), booleans, and integers. One column, language, contains 8 null values.

In [31]:
streamers.describe().round(2)

,total_followers,avg_concurrent_viewers
count,76.00,76.00
mean,3300885.76,31227.99
std,3793772.48,38307.22
min,-46964.00,634.00
25%,780000.00,6633.25
50%,2000000.00,14762.50
75%,3875000.00,36840.75
max,18700000.00,156401.00


The average amount of followers for this group of streamers is 3,300,886 and the average amount of concurrent viewers is 31,227. It looks like the minimum amount of followers is -46964, and I assume that negative numbers in this column are data entry issues meant to be positive since it's impossible to have a negative following count. The most followers any streamer in this list has is 18,700,000. The maximum amout of average viewers by a streamer is 156,401 and the minimum average viewers is 634.

**Streams**

In [12]:
streams.head()

,stream_id,streamer_id,started_at,ended_at,stream_duration_hrs,category,peak_viewers,title_has_hype_word,was_raid
0,SM00001,STR0001,2024-02-14 18:00:00,2024-02-14 21:00:00,3.0,Just Chatting,108989.0,False,True
1,SM00002,STR0001,2024-01-28 18:00:00,2024-01-28 21:00:00,3.0,Just Chatting,51635.0,True,False
2,SM00003,STR0001,2024-01-31 09:00:00,2024-01-31 16:00:00,7.0,Just Chatting,212127.0,False,False
3,SM00004,STR0001,2024-03-03 14:00:00,2024-03-03 21:00:00,7.0,Just Chatting,186745.0,False,True
4,SM00005,STR0001,2024-01-24 18:00:00,2024-01-24 21:00:00,3.0,Just Chatting,87723.0,False,True


In [21]:
streams.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1199 entries, 0 to 1198
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   stream_id            1199 non-null   object 
 1   streamer_id          1199 non-null   object 
 2   started_at           1199 non-null   object 
 3   ended_at             1199 non-null   object 
 4   stream_duration_hrs  1199 non-null   float64
 5   category             1149 non-null   object 
 6   peak_viewers         1119 non-null   float64
 7   title_has_hype_word  1199 non-null   bool   
 8   was_raid             1199 non-null   bool   
dtypes: bool(2), float64(2), object(5)
memory usage: 68.0+ KB


In [25]:
streams.describe()

,stream_duration_hrs,peak_viewers
count,1199.000000,1119.000000
mean,4.625104,38375.658624
std,7.995028,50377.085406
min,1.500000,-50.000000
25%,2.500000,7598.500000
50%,3.000000,20224.000000
75%,5.000000,47941.000000
max,96.000000,370165.000000


**Viewers**

In [13]:
viewers.head()

,viewer_id,age_group,country,account_age_days,subscription_tier,preferred_category
0,VWR00001,18-24,AR,355,free,IRL
1,VWR00002,25-34,CA,1670,free,Creative
2,VWR00003,35-44,US,1108,free,Just Chatting
3,VWR00004,45+,NaN,822,tier1,FPS
4,VWR00005,25-34,DE,2226,free,Just Chatting


In [22]:
viewers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   viewer_id           1000 non-null   object
 1   age_group           1000 non-null   object
 2   country             920 non-null    object
 3   account_age_days    1000 non-null   int64 
 4   subscription_tier   1000 non-null   object
 5   preferred_category  1000 non-null   object
dtypes: int64(1), object(5)
memory usage: 47.0+ KB


In [26]:
viewers.describe()

,account_age_days
count,1000.000000
mean,1881.119000
std,1036.144656
min,33.000000
25%,984.500000
50%,1966.500000
75%,2761.250000
max,3645.000000


### Task 1 Summary Report

Write your report here, summarizing the work done in the cells above.

# Task 2: Data Cleaning, Diagnostics, and Corrections

Identify and resolve issues in the dataset. These can be of different kinds, such as:
- Look for duplicates or implausible values (e.g., users with negative followers, invalid age groups). Thoroughly document your exploration of the data. Decide whether these are outliers or the result of a systemic problem. If the latter, resolve the problem here; if they are outliers, we will address those later.  
- Define at least three heuristics (common-sense rules) to flag unusual entries (e.g., someone watching dozens of streams per day). Discuss whether what you found are errors or artifacts.  

I will repeat the instructions here, but this is expected for each task: each code cell should focus on a specific analysis. You should run the code, and in the markdown below each code cell and its result, provide a detailed interpretation of what you have uncovered in that step. After completing the analyses, summarize the key findings in the indicated summary report cell below.


In [4]:
# start code for Task 2 here

### Task 2 summary report

# Task 3: Descriptive Statistics and Visualizations

- Create summary statistics and visualizations to understand user characteristics  
- Create summary statistics for streams data overall, and summaries by age group, account tire, or peak viewers  
- Visualize the distribution of streams durations, popular categories,   
- Show category-specific information and compare streamers based on categories  

In [5]:
#start code for Task 3 here

### Task 3 Summary report
summarize your findings here.

# Task 4: Dealing with Outliers and Missing Values

- Are there any missing values or outliers in the original data? Perform the analysis and report your findings.  
- Are there any outliers or missing values in the derived data? Perform the analysis and report your findings.  
- If there are no outliers or missing values, introduce 3–5% of missing values in a selected column and 1–2% of outliers in another column.  
- Experiment with two strategies for imputing missing values and two strategies for handling outliers. Show how each strategy affects the characteristics of the data.


In [6]:
#start code for Task 4 here

### Task 4 Summary report

# Task 5: Correlation and Behavioral Patterns

- Compute pairwise correlations between user features and the number of check-ins  
- Visualize the results using a correlation matrix (use a coolwarm color scheme and proper scaling)  
- Try Spearman rank correlation and explain when it is more appropriate than Pearson

In [7]:
# tast 5 code starts here

### Task 5 Summary report

# Task 6: Compare Behavior Across User Segments

- Segment users into 2–3 groups (e.g., by age bracket, account tire)  
- Compare their viewing behavior over time using boxplots or dot plots  
- Identify any outliers in group behavior and investigate them

In [8]:
# task 6 code start here

### Task 6 Summary report

# Task 7: Your Own EDA Questions

Formulate and explore two hypotheses of your own:  
- Propose questions based on curiosity (e.g., “Does a user's variety of categories influence morning vs. evening or session duration?”)  
- Use summary statistics and visualizations to answer each question  
- Briefly discuss what additional data would help you answer the questions more effectively


In [9]:
# task 7 code start here

### Task 7 Summary report

# Task 8: EDA Presentation (Insight Report)

Summarize your findings in a polished format:  
- Create 3–4 slides or a single-page report highlighting your most important insights  
- Include at least 2 charts and interpret them using clear, plain language  
- Apply Tufte’s principles: minimize chartjunk, maximize data-to-ink ratio, and ensure clear labeling  